neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
 
     def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
         """
-        Whether or not we should resume the pipeline (if the checkpoint exists)
+        Returns True if the checkpoint exists for each data inputs.
 
         :param context: execution context
-        :param data_container: data to resume
-        :return:
-        """
-        self.set_checkpoint_path(context.get_path())
-        return self._checkpoint_exists(data_container)
-
-    def _checkpoint_exists(self, data_container: DataContainer) -> bool:
-        """
-        Returns True if the checkpoints for each data input id exists
         :param data_container:
         :return:
         """
         for current_id in data_container.current_ids:
-            if not os.path.exists(self.get_checkpoint_file_path(current_id)):
+            if not os.path.exists(
+                    self.data_checkpointer.get_checkpoint_file_path(
+                        context.get_path(),
+                        current_id
+                    )
+            ):
                 return False
 
         return True
 
-    def get_checkpoint_file_path(self, current_id) -> str:
+
+class Checkpoint(ResumableMixin, BaseStep):
+    """
+    Resumable Checkpoint Step to load, and save both data checkpoints, and step checkpoints.
+    Checkpoint uses a list of step checkpointers(List[StepCheckpointer]), and data checkpointers(List[BaseCheckpointer]).
+
+    Data Checkpoints save the state of the data container (transformed data inputs, and expected outputs)
+    for the current execution mode (fit, fit_transform, or transform).
+
+    Step Checkpoints save the state of the fitted steps before the checkpoint
+    for the current execution mode (fit or fit_transform).
+
+    By default(no arguments specified), the Checkpoint step saves the step checkpoints for any fit or fit transform,
+    and saves a different data checkpoint with pickle data container checkpointers :
+
+    ```
+    Checkpoint(
+        step_checkpointers=[
+            StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+        ]
+        data_checkpointers=[
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.TRANSFORM),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT_TRANSFORM),
+        ]
+    )
+
+    # this is equivalent to :
+
+    Checkpoint()
+    ```
+    """
+    def __init__(
+            self,
+            step_checkpointers: List[StepCheckpointer] = None,
+            data_checkpointers: List[BaseCheckpointer] = None
+    ):
+        BaseStep.__init__(self)
+
+        if step_checkpointers is None:
+            step_checkpointers = [
+                StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+            ]
+        self.step_checkpointers = step_checkpointers
+
+        if data_checkpointers is None:
```

Comment:
Note that I don't really like this `if None` behavior. I'd create subclasses with each their responsibilitles. E.g.: 

```
class PickleDataCheckpoint(Checkpoint): 

    def __init__(self): 
        #. super...()
        self.data_checkpointers: List[BaseCheckpointer] = [
                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.TRANSFORM),
                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT),
                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT_TRANSFORM),
            ]
```

Same would go for the class `SaveStepAndPickleDataCheckpoint`, `SaveStepCheckpoint`, etc. 

This way if the user creates a `Checkpoint()` class but passes absolutely no checkpointer, then the checkpoint does nothing at all and prints a warning to tell the coder he's probably mistaken.